packages/teleterm/src/ui/DocumentCluster/ClusterResources/Kubes/useKubes.ts

```diff
@@ -14,18 +14,63 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+import { AttemptStatus } from 'shared/hooks/useAsync';
+
+import { Kube, ServerSideParams } from 'teleterm/services/tshd/types';
+import { useAppContext } from 'teleterm/ui/appContextProvider';
 import { useClusterContext } from 'teleterm/ui/DocumentCluster/clusterContext';
 
+import { useServerSideResources } from '../useServerSideResources';
+
+function getEmptyTableText(status: AttemptStatus) {
+  switch (status) {
+    case 'error':
+      return 'Failed to fetch kubernetes clusters.';
+    case '':
+      return 'Searching…';
+    case 'processing':
+      return 'Searching…';
+    case 'success':
+      return 'No kubernetes clusters found.';
+  }
+}
+
 export function useKubes() {
+  const appContext = useAppContext();
   const ctx = useClusterContext();
-  const kubes = ctx.getKubes();
-  const syncStatus = ctx.getSyncStatus().kubes;
+  const {
+    fetchAttempt,
+    updateQuery,
+    agentFilter,
+    prevPage,
+    nextPage,
+    updateSearch,
+    onAgentLabelClick,
+    updateSort,
+    pageCount,
+  } = useServerSideResources<Kube>(
+    { fieldName: 'name', dir: 'ASC' }, // default sort
+    (params: ServerSideParams) => appContext.resourcesService.fetchKubes(params)
+  );
 
   return {
-    pageSize: 100,
-    kubes,
-    syncStatus,
     connect: ctx.connectKube,
+    fetchAttempt,
+    kubes: fetchAttempt.data?.agentsList,
+    agentFilter,
+    updateQuery,
+    updateSearch,
+    onAgentLabelClick,
+    pageCount,
+    disabledRows: fetchAttempt.status === 'processing',
+    nextPage,
+    prevPage,
+    emptyTableText: getEmptyTableText(fetchAttempt.status),
+    customSort: {
+      dir: agentFilter.sort?.dir,
+      fieldName: agentFilter.sort?.fieldName,
+      onSort: updateSort,
```

Comment:
As I mentioned above, this is repeated three times across the hooks, I'd say let's move this into `useServerSideResources`.